includes/Classifai/Providers/AWS/AmazonPolly.php

```diff
@@ -0,0 +1,555 @@
+
+		return $new_settings;
+	}
+
+	/**
+	 * Connects to the Amazon Polly service.
+	 *
+	 * @param array $args Overridable args.
+	 * @return array
+	 */
+	public function connect_to_service( array $args = array() ): array {
+		$settings = $this->feature_instance->get_settings( static::ID );
+
+		$default = array(
+			'access_key_id'     => $settings[ static::ID ]['access_key_id'] ?? '',
+			'secret_access_key' => $settings[ static::ID ]['secret_access_key'] ?? '',
+			'aws_region'        => $settings[ static::ID ]['aws_region'] ?? 'us-east-1',
+		);
+
+		$default = wp_parse_args( $args, $default );
+
+		// Return if credentials don't exist.
+		if ( empty( $default['access_key_id'] ) || empty( $default['secret_access_key'] ) ) {
+			return array();
+		}
+
+		try {
+			$polly_client = $this->get_polly_client( $args );
+			$polly_voices = $polly_client->describeVoices();
+			return $polly_voices->get( 'Voices' );
+		} catch ( \Exception $e ) {
+			add_settings_error(
+				$this->feature_instance->get_option_name(),
+				'aws-polly-auth-failed',
+				esc_html__( 'Connection to Amazon Polly failed.', 'classifai' ),
+				'error'
+			);
+			return array();
+		}
+	}
+
+	/**
+	 * Returns HTML select dropdown options for voices.
+	 *
+	 * @param string $engine Engine type.
+	 * @return array
+	 */
+	public function get_voices_select_options( string $engine = '' ): array {
+		$settings = $this->feature_instance->get_settings( static::ID );
+		$voices   = $settings['voices'];
+		$options  = array();
+
+		if ( false === $voices ) {
+			return $options;
+		}
+
+		foreach ( $voices as $voice ) {
+			if (
+				! is_array( $voice ) ||
+				empty( $voice ) ||
+				(
+					! empty( $engine ) &&
+					! in_array( $engine, $voice['SupportedEngines'], true )
+				)
+			) {
+				continue;
+			}
+
+			$options[ $voice['Id'] ] = sprintf(
+				'%1$s - %2$s (%3$s)',
+				esc_html( $voice['LanguageName'] ),
+				esc_html( $voice['Name'] ),
+				esc_html( $voice['Gender'] )
+			);
+		}
+
+		// Sort the options.
+		asort( $options );
+
+		return $options;
+	}
+
+	/**
+	 * Synthesizes speech from a post item.
+	 *
+	 * @param int $post_id Post ID.
+	 * @return string|WP_Error
+	 */
+	public function synthesize_speech( int $post_id ) {
+		if ( empty( $post_id ) ) {
+			return new WP_Error(
+				'aws_polly_post_id_missing',
+				esc_html__( 'Post ID missing.', 'classifai' )
+			);
+		}
+
+		// We skip the user cap check if running under WP-CLI.
+		if ( ! current_user_can( 'edit_post', $post_id ) && ( ! defined( 'WP_CLI' ) || ! WP_CLI ) ) {
+			return new WP_Error(
+				'aws_polly_user_not_authorized',
+				esc_html__( 'Unauthorized user.', 'classifai' )
+			);
+		}
+
+		$normalizer          = new Normalizer();
+		$feature             = new TextToSpeech();
+		$settings            = $feature->get_settings();
+		$post                = get_post( $post_id );
+		$post_content        = $normalizer->normalize_content( $post->post_content, $post->post_title, $post_id );
```

Comment:
Looking at the [API docs](https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-polly-2016-06-10.html#synthesizespeech), seems there's a limit on the length of content we can send. Wondering if we should have a check here that we can return early with a proper error message if the content length is too long or if the error that will be returned in the `try/catch` block below is good enough?

Reply:
Yes, nice catch! However, while checking on this, I found that there is no error handling (in terms of displaying error messages) in place for the text-to-speech feature. I will investigate further on this.

Reply:
@dkotter This is added and PR is not ready for the review.